File: Leetcode206.java
public class Leetcode206 {
    public ListNode reverseList(ListNode head) {
        if(head==null || head.next == null){return head;}

        // Think like reversing the 'pointers'
        // Change the directions of pointers
        // last saves the remaining values when the link is deleted
        ListNode prev = null, curr = head, last = head.next;

        while(last != null){
            // reverse pointer
            curr.next = prev;
            // move prev and curr
            prev = curr;
            curr = last;
            last = last.next;
        }

        curr.next = prev;
        return curr;
    }
}

File: Python/Leetcode206.py
from contextlib import nullcontext
from email import header


class Leetcode206:
    def reverseList(self, head):
        if not head or not head.next: return head

        prev = None
        front = head
        back = head.next

        while back:
            # Set pointer
            front.next = prev

            # move pointers
            prev = front
            front = back
            back = back.next
        
        front.next = prev

        return front

